Install module tabs automatically

Quetzacoalt91 opened the pull request:
| Questions | Answers |
| --- | --- |
| Branch? | develop |
| Description? | This PR follows-up #3726 and will install all the tabs defined by a module automatically. The module developer should not have to install them himself anymore. This PR should not throw backward compatibility issues, because we do not add functions to the legacy `Module` class. |
| Type? | new feature |
| Category? | CO |
| BC breaks? | Nope |
| Deprecations? | Nope |
| Fixed ticket? | / |
| How to test? | When a module has an attribute `$tabs`, they are also installed after the module. |

TO DO:
- [x] Add a tab structure check
- [x] ~~Move tab CRUD requests in Doctrine class ?~~
- [x] Tests !
- [x] Module Events

### Example of tabs to register

```php
    protected $tabs = array(
        array(
            'name' => 'Merchant Expertise',
            'class_name' => 'AdminGamification',
            'parent_class' => 'AdminAdmin',
        ),
    );
```

### Documentation Draft, under construction

:page_facing_up:  https://docs.google.com/document/d/13cZ8emKULMi0vycZB1tIe0QZp6dfF-R1Hn-EMmHuLnQ

## Examples of "before/after" code
### dashgoals module

Before:
```php
	public function install()
	{
		Configuration::updateValue('PS_DASHGOALS_CURRENT_YEAR', date('Y'));
		for ($month = '01'; $month <= 12; $month = sprintf('%02d', $month + 1))
		{
			$key = Tools::strtoupper('dashgoals_traffic_'.$month.'_'.date('Y'));
			if (!ConfigurationKPI::get($key))
				ConfigurationKPI::updateValue($key, 600);
			$key = Tools::strtoupper('dashgoals_conversion_'.$month.'_'.date('Y'));
			if (!ConfigurationKPI::get($key))
				ConfigurationKPI::updateValue($key, 2);
			$key = Tools::strtoupper('dashgoals_avg_cart_value_'.$month.'_'.date('Y'));
			if (!ConfigurationKPI::get($key))
				ConfigurationKPI::updateValue($key, 80);
		}

		// Prepare tab
		$tab = new Tab();
		$tab->active = 1;
		$tab->class_name = 'AdminDashgoals';
		$tab->name = array();
		foreach (Language::getLanguages(true) as $lang)
			$tab->name[$lang['id_lang']] = 'Dashgoals';
		$tab->id_parent = -1;
		$tab->module = $this->name;

		return (
			$tab->add()
			&& parent::install()
			&& $this->registerHook('dashboardZoneTwo')
			&& $this->registerHook('dashboardData')
			&& $this->registerHook('actionAdminControllerSetMedia')
		);
	}

	public function uninstall()
	{
		$id_tab = (int)Tab::getIdFromClassName('AdminDashgoals');
		if ($id_tab)
		{
			$tab = new Tab($id_tab);
			$tab->delete();
		}

		return parent::uninstall();
	}
```

After: 
```php
	public function install()
	{
		Configuration::updateValue('PS_DASHGOALS_CURRENT_YEAR', date('Y'));
		for ($month = '01'; $month <= 12; $month = sprintf('%02d', $month + 1))
		{
			$key = Tools::strtoupper('dashgoals_traffic_'.$month.'_'.date('Y'));
			if (!ConfigurationKPI::get($key))
				ConfigurationKPI::updateValue($key, 600);
			$key = Tools::strtoupper('dashgoals_conversion_'.$month.'_'.date('Y'));
			if (!ConfigurationKPI::get($key))
				ConfigurationKPI::updateValue($key, 2);
			$key = Tools::strtoupper('dashgoals_avg_cart_value_'.$month.'_'.date('Y'));
			if (!ConfigurationKPI::get($key))
				ConfigurationKPI::updateValue($key, 80);
		}

		// We do not want to display the tab in the menu. Before, we needed to register the
		// ModuleAdminController in order to call it, but from now the Core will do it for us.

		return (
			parent::install()
			&& $this->registerHook('dashboardZoneTwo')
			&& $this->registerHook('dashboardData')
			&& $this->registerHook('actionAdminControllerSetMedia')
		);
	}

	// We do not need anything from uninstall() anymore, and can be safely removed. The core will do the job for the module developer
```

tiennou committed:
CO: DRY module tab registration

This allows module to declare their tabs in an instance variable and have them automatically added/removed on module installation/uninstallation.
It's also mostly-atomic, so you don't get leftover tabs in case of failure.

tiennou committed:
CO: Cleanup module overrides on a failed installation

Quetzacoalt91 committed:
CO: Move tab management in adapter to avoid potential retro-compatibility issues

kpodemski commented:
:+1: 

very nice, please remember to document this addition because it's something really, really cool and useful

kpodemski commented:
@Quetzacoalt91 what you think about adding something like 

`protected $sqlSchemaFile = null;`

if we have one it looks like:

`$sqlSchemaFile = 'init/schema.sql';`

and if file_exists it looks like:

```
$sql = array();
$sql['some_table_name'] = '
// sql to create some_table_name
';
$sql['other_table'] = '
// sql to create other_table
';
```

we're running loop on those queries and create database schema if this var is not null and file exists, also `onUninstall` we're deleting every table name by array_keys. 

What do you think? I can create PR for this, just let me know if you think it is a good addition, in my opinion it is because very often we need to write this few lines to install db schema from a file.

Quetzacoalt91 commented:
I like the idea ! However, I would prefer to deal with migration files instead of files list to declare on install or uninstall.
Today, we have some issues with modules which see their dependencies (sql schema, hooks, etc) installed or updated. The main reason is the developer forgot to add an upgrade file with the new module version.

This is a wide subject to handle. I agree with you that module developers has to write a lot of lines of code to make their module properly installed, but currently I do not have the answer to make all the dependencies up-to-date on each module version. Fortunately, the new module manager will help us. :wink:

Quetzacoalt91 committed:
CO: Update constructor of ModuleDataUpdater

pilothuy committed:
BO: Add icon property for Tab

Module can create new tab with icon in BO

Quetzacoalt91 committed:
Co: Fix wrong class sent to ModuleDataUpdater constructor

Quetzacoalt91 committed:
TE: Add stubs for ModuleUpdater

Quetzacoalt91 commented:
Tests now pass, time to improve this functionality with some doctrine entity.

Quetzacoalt91 committed:
CO: Introduce legacy Tab as entity

Quetzacoalt91 committed:
IN: Add missing translator for error management

kpodemski commented:
Yay!

Quetzacoalt91 committed:
CO: Introduce module tab (de)registration handlers

Quetzacoalt91 committed:
TE: Remove unused paremeter in ModuleRepository contruct

kpodemski commented:
I really hope that this would be done and merged into 1.7.1! :)

Quetzacoalt91 commented:
Hello @kpodemski,

I hope so too! I still working on it 😉 
I plan to go further, by adding automatically all the ModuleAdminControllers included with the module.
The developer should not have to declare them manually, in particular if he does not want to display them in the side menu.

Quetzacoalt91 committed:
CO: Add sf event on module actions, used for tabs

mickaelandrieu commented on src/Adapter/Module/Tab/ModuleTabManagementSubscriber.php:
```diff
@@ -0,0 +1,70 @@
+    
+    public function onModuleUninstall(ModuleManagementEvent $event)
+    {
+        $this->moduleTabUnregister->uninstallTabs($event->getModule());
```
why not `unregisterTabs` ?

Quetzacoalt91 replied:
Because I was still looking for the class name. This will be updated.

mickaelandrieu commented on src/Adapter/Module/Tab/ModuleTabUnregister.php:
```diff
@@ -24,15 +24,15 @@
  * International Registered Trademark & Property of PrestaShop SA
  */
 
+namespace PrestaShop\PrestaShop\Adapter\Module\Tab;
```
I'm not sure this should be located in `Adapter`

Quetzacoalt91 replied:
I put these classes in `Adapter` because I'm still using the Legacy `Tab` class 😞

mickaelandrieu commented on src/Core/Addon/Module/ModuleManager.php:
```diff
@@ -71,15 +73,24 @@ class ModuleManager implements AddonManagerInterface
         TranslatorInterface $translator,
-        Employee $employee = null)
+        Employee $employee = null,
+        EventDispatcherInterface $dispatcher = null)
```
In witch cases the dispatcher should be null ? I like a lot your PR but I need to be sure to rely on events in Symfony actions :)

Quetzacoalt91 replied:
This could happen when we do not instantiate this class with the service container from Symfony (i.e with `ModuleManagerBuider`:/).

And for the tests, I do not need the dispatcher to test this class. `TabRegister`and `TabUnregister` will have their own tests.

mickaelandrieu replied:
You can create a ``NullDispatcher`` that implement the interface and check that events are dispatched or simply [mock the interface](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Form/Tests/FormConfigTest.php#L74)

The real issue is still to rely on same class on both legacy and symfony parts :/

mickaelandrieu commented on src/Core/Addon/Module/ModuleManager.php:
```diff
@@ -246,8 +271,12 @@ public function disable($name)
         }
+        
+        if ($this->dispatcher) {
+            $this->dispatcher->dispatch(ModuleManagementEvent::DISABLE, new ModuleManagementEvent($module));
```
You may refactor this using a function 

```php
private function dispatch($eventName, Module $module){
    if ($this->dispatcher) { /.../ } 
}
```

Quetzacoalt91 replied:
Roger that!

mickaelandrieu commented on src/PrestaShopBundle/Resources/config/admin/services.yml:
```diff
@@ -206,4 +231,4 @@ services:
         properties:
             cacheDir: "%kernel.cache_dir%"
-            exportDir: "%kernel.cache_dir%/export"
+            exportDir: "%kernel.cache_dir%/export"
```
it's matter of taste, but add an extra line at the end of file :)

mickaelandrieu reviewed (changes requested):
This is a good job :)

Quetzacoalt91 committed:
CO: Code review on module tabs managers

Quetzacoalt91 committed:
TE: Add tests for ModuleTabRegister

Quetzacoalt91 committed:
BO: Add dispatcher to trigger tab (un)registration

Quetzacoalt91 committed:
BO: Install all ModuleAdminControllers, even if undeclared by module

Quetzacoalt91 committed:
TE: Implement test to check auto-registered ModuleAdminControllers

Quetzacoalt91 commented:
This improvement is done and can be reviewed. 🙂

Quetzacoalt91 committed:
CO: Module tab registration stability improvements

aleeks commented on src/Adapter/Module/Tab/ModuleTabRegister.php:
```diff
@@ -118,16 +119,17 @@ protected function addUndeclaredTabs($moduleName, array $tabs)
             );
         }
-        
+        dump($tabs);
```
?!!! :rage:

Quetzacoalt91 replied:
Woups, removed

Quetzacoalt91 commented on src/Adapter/Module/Tab/ModuleTabRegister.php:
```diff
@@ -0,0 +1,237 @@
+            }
+            
+            $tabs[] = array(
+                'class_name' => $adminControllerName,
```
To be added:
```php
'visible' => false,
``` 

ModuleAdminController are not going to be visible. We just want them to be callable for display in the BackOffice

Quetzacoalt91 commented on src/Adapter/Module/Tab/ModuleTabRegister.php:
```diff
@@ -0,0 +1,237 @@
+        // won't be directly linked to the tab creation
+        // @ToDo
+        $tab = new Tab();
+        $tab->active = $data->getBoolean('visible', false);
```
The tabs explicitly declared by the module developer are most likely to be displayed. This property needs to be set to `true`

Quetzacoalt91 committed:
CO: Change default values of tab registration

Quetzacoalt91 committed:
CO: Handle (non-)translated names for tabs

Quetzacoalt91 committed:
CO: Add event NullDispatcher and remove tests in ModuleManager

Quetzacoalt91 commented:
Just added `NullDispatcher` class regarding @mickaelandrieu code review, and added function to handle translated or non-translated names for the tab to register.

maximebiloe commented:
Let's merge it! 🙈

Thank you @Quetzacoalt91

kpodemski commented:
![image](https://cloud.githubusercontent.com/assets/2137763/22518011/59dcf700-e8ab-11e6-865e-084d76b6b583.png)

kpodemski commented:
Hey @Quetzacoalt91 

please remind someone to put this in docs! :)

Quetzacoalt91 commented:
True, we discussed about that documentation with @xBorderie last week :)
